File: packages/i18nify-go/modules/currency/convert_currency.go
package currency

import (
	"fmt"
	"math"
	"strconv"
)

// ConvertCurrency converts an amount from one currency to another using the provided exchange rate.
// The result is rounded to the number of decimal places defined by toCurrency's minor unit.
//
// Example:
//
//	ConvertCurrency(100.0, "USD", "JPY", 149.50)  // → 14950, nil  (JPY: 0 decimal places)
//	ConvertCurrency(100.0, "USD", "KWD", 0.307)   // → 30.7, nil   (KWD: 3 decimal places)
func ConvertCurrency(amount float64, fromCurrency string, toCurrency string, exchangeRate float64) (float64, error) {
	if fromCurrency == "" {
		return 0, fmt.Errorf("fromCurrency cannot be empty")
	}
	if toCurrency == "" {
		return 0, fmt.Errorf("toCurrency cannot be empty")
	}
	if exchangeRate <= 0 {
		return 0, fmt.Errorf("exchangeRate must be a positive number, got %v", exchangeRate)
	}

	if _, err := GetCurrencyInformation(fromCurrency); err != nil {
		return 0, fmt.Errorf("invalid fromCurrency: %v", err)
	}

	toInfo, err := GetCurrencyInformation(toCurrency)
	if err != nil {
		return 0, fmt.Errorf("invalid toCurrency: %v", err)
	}

	toMinorUnit, err := strconv.Atoi(toInfo.MinorUnit)
	if err != nil {
		return 0, fmt.Errorf("invalid minor unit for currency '%s': %v", toCurrency, err)
	}

	multiplier := math.Pow(10, float64(toMinorUnit))
	return math.Round(amount*exchangeRate*multiplier) / multiplier, nil
}

File: packages/i18nify-js/src/modules/currency/convertCurrency.ts
import { withErrorBoundary } from '../../common/errorBoundary';
import { CurrencyCodeType } from './types';
import CURRENCY_INFO from './data/currencyConfig.json';

const convertCurrency = (
  amount: number,
  options: {
    fromCurrency: CurrencyCodeType;
    toCurrency: CurrencyCodeType;
    exchangeRate: number;
  },
): number => {
  const { fromCurrency, toCurrency, exchangeRate } = options;

  if (!fromCurrency || !(fromCurrency in CURRENCY_INFO))
    throw new Error(
      `The provided source currency code is invalid. The received value was: ${String(fromCurrency)}. Please ensure you pass a valid ISO 4217 currency code.`,
    );

  if (!toCurrency || !(toCurrency in CURRENCY_INFO))
    throw new Error(
      `The provided target currency code is invalid. The received value was: ${String(toCurrency)}. Please ensure you pass a valid ISO 4217 currency code.`,
    );

  if (
    typeof exchangeRate !== 'number' ||
    !isFinite(exchangeRate) ||
    exchangeRate <= 0
  )
    throw new Error(
      `The exchange rate must be a positive finite number. The received value was: ${String(exchangeRate)}.`,
    );

  const toMinorUnit = Number(CURRENCY_INFO[toCurrency].minor_unit);
  const multiplier = Math.pow(10, toMinorUnit);

  return Math.round(amount * exchangeRate * multiplier) / multiplier;
};

export default withErrorBoundary<typeof convertCurrency>(convertCurrency);
